### PR TITLE
[FLINK-11898] [table-planner-blink] Support code generation for all Blink built-in functions and operators

### PR DESCRIPTION
**Pull request by wuchong**
## What is the purpose of the change

Support code generation for all Blink built-in functions and operators.

## Brief change log

changes to `flink-table-planner-blink`

 - Add `BinaryStringCallGen` include `generate...` methods for BinaryString.
 - Add `ScalarFunctionGens` for generate function using `BuiltinMethods`.
 - Register all the Blink built-in SqlFunctions to `FlinkSqlOperatorTable`

changes to `flink-table-runtime-blink`
 - Add `SqlDateTimeUtils` to place all the date time related methods.
 - Add `SqlFunctionUtils` to place all the other builtin methods.
 - Add Calcite Avatica dependency to `flink-table-runtime-blink` because we need Avatica's `TimeUnitRange` and `DateTimeUtils` when in runtime `SqlDateTimeUtils`.

## Verifying this change

- We add a lot of expression tests under `org/apache/flink/table/expressions` to test built-in functions. 
- Move all the IT tests in `BuiltinScalarFunctionITCase` (in blink repository) to `ScalarFunctionsTest` unit tests to save test time.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/util/SegmentsUtil.java [hunk 1]
```diff
@@ -987,4 +1010,82 @@ private static void setTwoByteSlowly(
...
+					return i;
+				}
+			}
+			return -1;
+		} else {
+			return findSlow(segments1, offset1, numBytes1, segments2, offset2, numBytes2);
+		}
+	}
+
+	private static int findSlow(
```
Comment (JingsongLi) [inline flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/util/SegmentsUtil.java]: findInMultiSegments?

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/util/SegmentsUtil.java [hunk 2]
```diff
@@ -987,4 +1010,82 @@ private static void setTwoByteSlowly(
...
+	/**
+	 * Copy bytes of segments to output view.
+	 * Note: It just copies the data in, not include the length.
+	 *
+	 * @param segments source segments
+	 * @param offset offset for segments
+	 * @param sizeInBytes size in bytes
+	 * @param target target output view
+	 */
+	public static void copyBytesToView(MemorySegment[] segments, int offset,
```
Comment (JingsongLi) [inline flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/util/SegmentsUtil.java]: `CopyToView` already exists

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/util/StringUtf8Utils.java [hunk 3]
```diff
@@ -0,0 +1,342 @@
...
+public class StringUtf8Utils {
+
+	private static final int MAX_CHARS_LENGTH = 1024 * 32;
+	private static final int MAX_BYTES_LENGTH = 1024 * 64;
+	public static final int MAX_BYTES_PER_CHAR = 3;
+
+	private static final ThreadLocal<char[]> CHARS_LOCAL = new ThreadLocal<>();
+	private static final ThreadLocal<byte[]> BYTES_LOCAL = new ThreadLocal<>();
+
+	public static char[] allocateChars(int length) {
```
Comment (JingsongLi) [inline flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/util/StringUtf8Utils.java]: Move allocateChars and allocateBytes(allocateReuseBytes existed) to SegmentsUtil

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/util/StringUtf8Utils.java [hunk 4]
```diff
@@ -0,0 +1,342 @@
...
+import java.io.UnsupportedEncodingException;
+import java.util.Arrays;
+
+/**
+ * String utf-8 utils.
+ *
+ * <p>{@code StringUtf8Utils} refers to the implementation from SerializeWriter and IOUtils of Alibaba fastjson.
+ * The difference is that StringUtf8Utils need to handle the wrong code, just like StringCoding.decode.
+ */
+public class StringUtf8Utils {
```
Comment (JingsongLi) [inline flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/util/StringUtf8Utils.java]: If you want to introduce `StringUtf8Utils`, please modify all `new String` and `String.getBytes`.
e.g.: `BinaryString.toString`.

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/dataformat/BinaryString.java [hunk 5]
```diff
@@ -242,4 +330,1508 @@ private int compareMultiSegments(BinaryString other) {
...
+
+	public static String safeToString(BinaryString str) {
+		if (str == null) {
+			return null;
+		} else {
+			return str.toString();
+		}
+	}
+
+	private boolean inOneSegment() {
```
Comment (JingsongLi) [inline flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/dataformat/BinaryString.java]: inFirstSegment?

**Comment by wuchong** [comment]
Thanks @JingsongLi  for reviewing this. I have addressed all the comments.

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/util/SegmentsUtil.java [hunk 6]
```diff
@@ -74,6 +76,42 @@
 		return bytes;
 	}
 
+	public static byte[] allocateBytes(int length) {
```
Comment (JingsongLi) [inline flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/util/SegmentsUtil.java]: use existing `allocateReuseBytes`.

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/codegen/CodeGenUtils.scala [hunk 7]
```diff
@@ -174,6 +178,86 @@ object CodeGenUtils {
       (classOf[BaseRow].isAssignableFrom(clazz) ||
           clazz == getInternalClassForType(createInternalTypeFromTypeInfo(t)))
 
+  def hashCodeForType(ctx: CodeGeneratorContext, t: InternalType, term: String): String = t match {
```
Comment (JingsongLi) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/codegen/CodeGenUtils.scala]: Merge `HashCodeGenerator`?

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/calcite/avatica/util/DateTimeUtils.java [hunk 8]
```diff
@@ -0,0 +1,1195 @@
...
+import java.text.NumberFormat;
+import java.text.ParsePosition;
+import java.text.SimpleDateFormat;
+import java.util.Calendar;
+import java.util.Date;
+import java.util.Locale;
+import java.util.TimeZone;
+
+/*
+ * THIS FILE HAS BEEN COPIED FROM THE APACHE CALCITE PROJECT UNTIL CALCITE-1884 IS FIXED.
```
Comment (KurtYoung) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/calcite/avatica/util/DateTimeUtils.java]: CALCITE-1884 is already fixed, do we still need this?

Comment (wuchong) [reply flink-table/flink-table-planner-blink/src/main/java/org/apache/calcite/avatica/util/DateTimeUtils.java]: It seems that CALCITE-1884 is not the root cause or not correctly fixed. I have tried to remove this file, but several date time cases were failed. I have create [FLINK-12063](https://issues.apache.org/jira/browse/FLINK-12063) ticket to fix this later.

**Comment by wuchong** [comment]
Thanks @JingsongLi  for the comments. I have addressed them.

**Comment by JingsongLi** [comment]
+1 LGTM

**Comment by KurtYoung** [comment]
also +1 from me

**Comment by wuchong** [comment]
Merging

**Pull request closed without merging**